Debug builds fail:

```
In file included from /Volumes/Data/worker/macOS-Sonoma-Debug-Build-EWS/build/Source/WebKit/Shared/Cocoa/CoreIPCNull.mm:27:
In file included from /Volumes/Data/worker/macOS-Sonoma-Debug-Build-EWS/build/Source/WebKit/Shared/Cocoa/CoreIPCNull.h:28:
In file included from /Volumes/Data/worker/macOS-Sonoma-Debug-Build-EWS/build/WebKitBuild/Debug/usr/local/include/wtf/RetainPtr.h:29:
In file included from /Volumes/Data/worker/macOS-Sonoma-Debug-Build-EWS/build/WebKitBuild/Debug/usr/local/include/wtf/HashTraits.h:29:
/Volumes/Data/worker/macOS-Sonoma-Debug-Build-EWS/build/WebKitBuild/Debug/usr/local/include/wtf/text/ASCIILiteral.h:146:48: error: unsafe buffer access [-Werror,-Wunsafe-buffer-usage]
        ASSERT_UNDER_CONSTEXPR_CONTEXT(isASCII(characters[i]));
                                               ^~~~~~~~~~
```

--- a/Source/JavaScriptCore/runtime/IntlCollator.cpp
+++ b/Source/JavaScriptCore/runtime/IntlCollator.cpp
@@ -89,7 +89,7 @@ Vector<String> IntlCollator::sortLocaleData(const String& locale, RelevantExtens
             int32_t length = 0;
             while ((pointer = uenum_next(enumeration.get(), &length, &status)) && U_SUCCESS(status)) {
                 // 10.2.3 "The values "standard" and "search" must not be used as elements in any [[sortLocaleData]][locale].co and [[searchLocaleData]][locale].co array."
-                String collation({ pointer, static_cast<size_t>(length) });
+                String collation(unsafeForgeSpan(pointer, static_cast<size_t>(length)));
                 if (collation == "standard"_s || collation == "search"_s)
                     continue;
                 if (auto mapped = mapICUCollationKeywordToBCP47(collation))

--- a/Source/JavaScriptCore/runtime/IntlDateTimeFormat.cpp
+++ b/Source/JavaScriptCore/runtime/IntlDateTimeFormat.cpp
@@ -156,7 +156,7 @@ Vector<String> IntlDateTimeFormat::localeData(const String& locale, RelevantExte
         int32_t nameLength;
         while (const char* availableName = uenum_next(calendars, &nameLength, &status)) {
             ASSERT(U_SUCCESS(status));
-            String calendar = String({ availableName, static_cast<size_t>(nameLength) });
+            String calendar = String(unsafeForgeSpan(availableName, static_cast<size_t>(nameLength)));
             keyLocaleData.append(calendar);
             // Adding "islamicc" candidate for backward compatibility.
             if (calendar == "islamic-civil"_s)

--- a/Source/JavaScriptCore/runtime/IntlLocale.cpp
+++ b/Source/JavaScriptCore/runtime/IntlLocale.cpp
@@ -598,7 +598,7 @@ JSArray* IntlLocale::calendars(JSGlobalObject* globalObject)
     const char* pointer;
     int32_t length = 0;
     while ((pointer = uenum_next(calendars.get(), &length, &status)) && U_SUCCESS(status)) {
-        String calendar({ pointer, static_cast<size_t>(length) });
+        String calendar(unsafeForgeSpan(pointer, static_cast<size_t>(length)));
         if (auto mapped = mapICUCalendarKeywordToBCP47(calendar))
             elements.append(WTFMove(mapped.value()));
         else
@@ -636,7 +636,7 @@ JSArray* IntlLocale::collations(JSGlobalObject* globalObject)
     const char* pointer;
     int32_t length = 0;
     while ((pointer = uenum_next(enumeration.get(), &length, &status)) && U_SUCCESS(status)) {
-        String collation({ pointer, static_cast<size_t>(length) });
+        String collation(unsafeForgeSpan(pointer, static_cast<size_t>(length)));
         // 1.1.3 step 4, The values "standard" and "search" must be excluded from list.
         if (collation == "standard"_s || collation == "search"_s)
             continue;

--- a/Source/JavaScriptCore/runtime/IntlObject.cpp
+++ b/Source/JavaScriptCore/runtime/IntlObject.cpp
@@ -1631,7 +1631,7 @@ const Vector<String>& intlAvailableCalendars()
             int32_t length = 0;
             const char* pointer = uenum_next(enumeration.get(), &length, &status);
             ASSERT(U_SUCCESS(status));
-            String calendar({ pointer, static_cast<size_t>(length) });
+            String calendar(unsafeForgeSpan(pointer, static_cast<size_t>(length)));
             if (auto mapped = mapICUCalendarKeywordToBCP47(calendar))
                 return createImmortalThreadSafeString(WTFMove(mapped.value()));
             return createImmortalThreadSafeString(WTFMove(calendar));
@@ -1702,7 +1702,7 @@ static JSArray* availableCollations(JSGlobalObject* globalObject)
             throwTypeError(globalObject, scope, "failed to enumerate available collations"_s);
             return { };
         }
-        String collation({ pointer, static_cast<size_t>(length) });
+        String collation(unsafeForgeSpan(pointer, static_cast<size_t>(length)));
         if (collation == "standard"_s || collation == "search"_s)
             continue;
         if (auto mapped = mapICUCollationKeywordToBCP47(collation))
@@ -1759,7 +1759,7 @@ static JSArray* availableCurrencies(JSGlobalObject* globalObject)
             throwTypeError(globalObject, scope, "failed to enumerate available currencies"_s);
             return { };
         }
-        String currency({ pointer, static_cast<size_t>(length) });
+        String currency(unsafeForgeSpan(pointer, static_cast<size_t>(length)));
         if (currency == "EQE"_s)
             continue;
         if (currency == "LSM"_s)
@@ -1868,7 +1868,7 @@ const Vector<String>& intlAvailableTimeZones()
             int32_t length = 0;
             const char* pointer = uenum_next(enumeration.get(), &length, &status);
             ASSERT(U_SUCCESS(status));
-            String timeZone({ pointer, static_cast<size_t>(length) });
+            String timeZone(unsafeForgeSpan(pointer, static_cast<size_t>(length)));
             if (isValidTimeZoneNameFromICUTimeZone(timeZone)) {
                 if (auto mapped = canonicalizeTimeZoneNameFromICUTimeZone(WTFMove(timeZone)))
                     temporary.append(WTFMove(mapped.value()));

--- a/Source/JavaScriptCore/runtime/IntlPluralRules.cpp
+++ b/Source/JavaScriptCore/runtime/IntlPluralRules.cpp
@@ -190,7 +190,7 @@ JSObject* IntlPluralRules::resolvedOptions(JSGlobalObject* globalObject) const
     unsigned index = 0;
     while (const char* result = uenum_next(keywords.get(), &resultLength, &status)) {
         ASSERT(U_SUCCESS(status));
-        categories->putDirectIndex(globalObject, index++, jsNontrivialString(vm, String({ result, static_cast<size_t>(resultLength) })));
+        categories->putDirectIndex(globalObject, index++, jsNontrivialString(vm, String(unsafeForgeSpan(result, static_cast<size_t>(resultLength)))));
         RETURN_IF_EXCEPTION(scope, { });
     }
     options->putDirect(vm, Identifier::fromString(vm, "pluralCategories"_s), categories);

--- a/Source/JavaScriptCore/runtime/JSONObject.cpp
+++ b/Source/JavaScriptCore/runtime/JSONObject.cpp
@@ -174,7 +174,7 @@ static inline String gap(JSGlobalObject* globalObject, JSValue space)
     VM& vm = globalObject->vm();
     auto scope = DECLARE_THROW_SCOPE(vm);
 
-    const unsigned maxGapLength = 10;
+    constexpr unsigned maxGapLength = 10;
     space = unwrapBoxedPrimitive(globalObject, space);
     RETURN_IF_EXCEPTION(scope, { });
 
@@ -191,7 +191,7 @@ static inline String gap(JSGlobalObject* globalObject, JSValue space)
         char spaces[maxGapLength];
         for (size_t i = 0; i < count; ++i)
             spaces[i] = ' ';
-        return String({ spaces, count });
+        return String(std::span { spaces }.first(count));
     }
 
     // If the space value is a string, use it as the gap string, otherwise use no gap string.

--- a/Source/JavaScriptCore/tools/FunctionAllowlist.cpp
+++ b/Source/JavaScriptCore/tools/FunctionAllowlist.cpp
@@ -70,7 +70,7 @@ FunctionAllowlist::FunctionAllowlist(const char* filename)
         if (!length)
             continue;
         
-        m_entries.add(String({ line, length }));
+        m_entries.add(String(unsafeForgeSpan(line, length)));
     }
 
     int result = fclose(f);

--- a/Source/JavaScriptCore/tools/FunctionOverrides.cpp
+++ b/Source/JavaScriptCore/tools/FunctionOverrides.cpp
@@ -218,13 +218,13 @@ static String parseClause(const char* keyword, size_t keywordLength, FILE* file,
     const char* delimiterEnd = strchr(delimiterStart, '{');
     if (!delimiterEnd)
         FAIL_WITH_ERROR(SYNTAX_ERROR, ("Missing { after '", keyword, "' clause start delimiter:\n", line, "\n"));
-    
+
     size_t delimiterLength = delimiterEnd - delimiterStart;
-    String delimiter({ delimiterStart, delimiterLength });
+    String delimiter(unsafeForgeSpan(delimiterStart, delimiterLength));
 
     if (hasDisallowedCharacters(delimiterStart, delimiterLength))
         FAIL_WITH_ERROR(SYNTAX_ERROR, ("Delimiter '", delimiter, "' cannot have '{', '}', or whitespace:\n", line, "\n"));
-    
+
     CString terminatorCString = makeString('}', delimiter).ascii();
     const char* terminator = terminatorCString.data();
     line = delimiterEnd; // Start from the {.

--- a/Source/WTF/wtf/StdLibExtras.h
+++ b/Source/WTF/wtf/StdLibExtras.h
@@ -762,11 +762,11 @@ template<typename OptionalType> auto valueOrDefault(OptionalType&& optionalValue
 #pragma GCC diagnostic push
 #pragma GCC diagnostic ignored "-Wcast-align"
 template<typename T, typename U, std::size_t Extent>
-std::span<T, Extent == std::dynamic_extent ? std::dynamic_extent : (sizeof(U) * Extent) / sizeof(T)> spanReinterpretCast(std::span<U, Extent> span)
+constexpr std::span<T, Extent == std::dynamic_extent ? std::dynamic_extent : (sizeof(U) * Extent) / sizeof(T)> spanReinterpretCast(std::span<U, Extent> span)
 {
     if constexpr (Extent == std::dynamic_extent) {
         if constexpr (sizeof(U) < sizeof(T) || sizeof(U) % sizeof(T))
-            RELEASE_ASSERT_WITH_MESSAGE(!(span.size_bytes() % sizeof(T)), "spanReinterpretCast will not change size in bytes from source");
+            RELEASE_ASSERT_UNDER_CONSTEXPR_CONTEXT(!(span.size_bytes() % sizeof(T))); // Refuse to change size in bytes from source.
     } else
         static_assert(!((sizeof(U) * Extent) % sizeof(T)), "spanReinterpretCast will not change size in bytes from source");
 
@@ -827,7 +827,7 @@ void memsetSpan(std::span<T, Extent> destination, uint8_t byte)
 // Use this when we can't edit the imported API and it doesn't offer
 // begin() / end() or a span accessor.
 template<typename T, std::size_t Extent = std::dynamic_extent>
-inline auto unsafeForgeSpan(T* ptr, size_t size)
+inline constexpr auto unsafeForgeSpan(T* ptr, size_t size)
 {
 WTF_ALLOW_UNSAFE_BUFFER_USAGE_BEGIN
     return std::span<T, Extent> { ptr, size };

--- a/Source/WTF/wtf/text/ASCIILiteral.h
+++ b/Source/WTF/wtf/text/ASCIILiteral.h
@@ -26,14 +26,11 @@
 
 #pragma once
 
-#include <wtf/Compiler.h>
-
-WTF_ALLOW_UNSAFE_BUFFER_USAGE_BEGIN
-
 #include <span>
 #include <string>
 #include <type_traits>
 #include <wtf/ASCIICType.h>
+#include <wtf/Compiler.h>
 #include <wtf/Forward.h>
 #include <wtf/HashFunctions.h>
 #include <wtf/StdLibExtras.h>
@@ -52,7 +49,7 @@ public:
     static constexpr ASCIILiteral fromLiteralUnsafe(const char* string)
     {
         ASSERT_UNDER_CONSTEXPR_CONTEXT(string);
-        return ASCIILiteral { string };
+        return ASCIILiteral { unsafeForgeSpan(string, std::char_traits<char>::length(string) + 1) };
     }
 
     WTF_EXPORT_PRIVATE void dump(PrintStream& out) const;
@@ -67,7 +64,7 @@ public:
 
     constexpr const char* characters() const { return m_charactersWithNullTerminator.data(); }
     constexpr size_t length() const { return !m_charactersWithNullTerminator.empty() ? m_charactersWithNullTerminator.size() - 1 : 0; }
-    std::span<const LChar> span8() const { return { bitwise_cast<const LChar*>(characters()), length() }; }
+    std::span<const LChar> span8() const { return spanReinterpretCast<const LChar>(m_charactersWithNullTerminator.first(length())); }
     std::span<const char> spanIncludingNullTerminator() const { return m_charactersWithNullTerminator; }
     size_t isEmpty() const { return m_charactersWithNullTerminator.size() <= 1; }
 
@@ -83,12 +80,12 @@ public:
     bool isDeletedValue() const { return characters() == reinterpret_cast<char*>(-1); }
 
 private:
-    constexpr explicit ASCIILiteral(const char* characters)
-        : m_charactersWithNullTerminator(characters, std::char_traits<char>::length(characters) + 1)
+    constexpr explicit ASCIILiteral(std::span<const char> spanWithNullTerminator)
+        : m_charactersWithNullTerminator(spanWithNullTerminator)
     {
 #if ASSERT_ENABLED
-    for (size_t i = 0; i < length(); ++i)
-        ASSERT_UNDER_CONSTEXPR_CONTEXT(isASCII(m_charactersWithNullTerminator[i]));
+        for (size_t i = 0, size = length(); i < size; ++i)
+            ASSERT_UNDER_CONSTEXPR_CONTEXT(isASCII(m_charactersWithNullTerminator[i]));
 #endif
     }
 
@@ -145,11 +142,12 @@ constexpr ASCIILiteral operator"" _s(const char* characters, size_t)
 
 constexpr std::span<const LChar> operator"" _span(const char* characters, size_t n)
 {
+    auto span = spanReinterpretCast<const LChar>(unsafeForgeSpan(characters, n));
 #if ASSERT_ENABLED
-    for (size_t i = 0; i < n; ++i)
-        ASSERT_UNDER_CONSTEXPR_CONTEXT(isASCII(characters[i]));
+    for (size_t i = 0, size = span.size(); i < size; ++i)
+        ASSERT_UNDER_CONSTEXPR_CONTEXT(isASCII(span[i]));
 #endif
-    return std::span { bitwise_cast<const LChar*>(characters), n };
+    return span;
 }
 
 } // inline StringLiterals
@@ -157,5 +155,3 @@ constexpr std::span<const LChar> operator"" _span(const char* characters, size_t
 } // namespace WTF
 
 using namespace WTF::StringLiterals;
-
-WTF_ALLOW_UNSAFE_BUFFER_USAGE_END

--- a/Source/WTF/wtf/text/SuperFastHash.h
+++ b/Source/WTF/wtf/text/SuperFastHash.h
@@ -1,5 +1,5 @@
 /*
- * Copyright (C) 2005-2023 Apple Inc. All rights reserved.
+ * Copyright (C) 2005-2024 Apple Inc. All rights reserved.
  * Copyright (C) 2010 Patrick Gansterer <paroga@paroga.com>
  *
  * This library is free software; you can redistribute it and/or
@@ -22,7 +22,10 @@
 #pragma once
 
 #include <span>
+#include <wtf/Compiler.h>
 #include <wtf/text/StringHasher.h>
+
+WTF_ALLOW_UNSAFE_BUFFER_USAGE_BEGIN
 
 namespace WTF {
 
@@ -273,5 +276,7 @@ private:
 };
 
 } // namespace WTF
+
+WTF_ALLOW_UNSAFE_BUFFER_USAGE_END
 
 using WTF::SuperFastHash;

--- a/Tools/WebKitTestRunner/TestController.cpp
+++ b/Tools/WebKitTestRunner/TestController.cpp
@@ -1,5 +1,5 @@
 /*
- * Copyright (C) 2010-2022 Apple Inc. All rights reserved.
+ * Copyright (C) 2010-2024 Apple Inc. All rights reserved.
  *
  * Redistribution and use in source and binary forms, with or without
  * modification, are permitted provided that the following conditions
@@ -1505,7 +1505,7 @@ WKURLRef TestController::createTestURL(const char* pathOrURL)
     if (length >= 7 && strstr(pathOrURL, "file://")) {
         auto url = adoptWK(WKURLCreateWithUTF8CString(pathOrURL));
         auto path = testPath(url.get());
-        if (!m_usingServerMode && !WTF::FileSystemImpl::fileExists(String({ path.c_str(), path.length() }))) {
+        if (!m_usingServerMode && !WTF::FileSystemImpl::fileExists(String(std::span { path }))) {
             printf("Failed: File for URL ‘%s’ was not found or is inaccessible\n", pathOrURL);
             return 0;
         }
@@ -1540,7 +1540,7 @@ WKURLRef TestController::createTestURL(const char* pathOrURL)
     auto cPath = buffer.get();
     auto url = adoptWK(WKURLCreateWithUTF8CString(cPath));
     auto path = testPath(url.get());
-    if (!m_usingServerMode && !WTF::FileSystemImpl::fileExists(String({ path.c_str(), path.length() }))) {
+    if (!m_usingServerMode && !WTF::FileSystemImpl::fileExists(String(std::span { path }))) {
         printf("Failed: File ‘%s’ was not found or is inaccessible\n", pathOrURL);
         return 0;
     }

